include/margo-monitoring.h

```diff
@@ -0,0 +1,397 @@
+/**
+ * @file margo-monitoring.h
+ *
+ * (C) 2022 The University of Chicago
+ *
+ * See COPYRIGHT in top-level directory.
+ */
+#ifndef __MARGO_MONITORING_H
+#define __MARGO_MONITORING_H
+
+#include <abt.h>
+#include <mercury.h>
+#include <mercury_types.h>
+#include <mercury_bulk.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+struct hg_bulk_attr;
+struct margo_instance;
+typedef struct margo_instance*       margo_instance_id;
+typedef struct margo_request_struct* margo_request;
+
+extern const struct margo_monitor* margo_default_monitor;
+
+/**
+ * The margo_monitor_data_t union is used in all the
+ * margo_monitor_*_args structures so that developers
+ * of a custom monitoring system can pass some data
+ * between a MARGO_MONITOR_FN_START event and its
+ * corresponding MARGO_MONITOR_FN_END.
+ */
+typedef union {
+    int64_t i;
+    double  f;
+    void*   p;
+} margo_monitor_data_t;
+
+typedef enum margo_monitor_event_t
+{
+    MARGO_MONITOR_FN_START,
+    MARGO_MONITOR_FN_END,
+    MARGO_MONITOR_POINT
```

Comment:
What is MARGO_MONITOR_POINT for?

Reply:
I initially thought that some callback would only be called once instead of twice, such as functions for which we don't really care about a duration (margo_register/deregister, margo_addr/bulk_free, etc.), but ended up systematically using START/END and never using POINT. Let me know if you think it would make sense to either use POINT for some of the functions, or to remove the MARGO_MONITOR_POINT value from the enum.

Reply:
Ok, I see.  I'm fine with it being left there with a comment then.